compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/ScopeUtils.kt

```diff
@@ -25,6 +27,25 @@ import org.jetbrains.kotlin.fir.types.*
 import org.jetbrains.kotlin.fir.types.impl.ConeClassLikeTypeImpl
 import org.jetbrains.kotlin.fir.types.impl.ConeTypeParameterTypeImpl
 import org.jetbrains.kotlin.name.ClassId
+import org.jetbrains.kotlin.types.SmartcastStability
+
+fun FirExpressionWithSmartcast.smartcastScope(
```

Comment:
So... `FirExpressionWithSmartcast.typeRef` returns either `smartcastType` or `originalType`, depending on stability. Then, can we simply do something like:
```
... = this.typeRef.coneType.scope(useSiteSession, scopeSession, ...)
```
?

Reply:
No, this doesn't work. For unstable smartcast, it's crucial that we use a scope that can serve candidates that are only available from the smartcast type. Otherwise, those candidates will be unresolved rather than resolved with "SMARTCAST_IMPOSSIBLE".